Améliore le comportement des archives de tuto/article

pierre-24 opened the pull request:
| Q | R |
| --- | --- |
| Correction de bugs ? | oui |
| Nouvelle Fonctionnalité ? | ~oui |
| Tickets concernés | #1458, #1460 |

Cette PR vise à régler deux choses fortement liées : les archives des tutoriels ou articles sont en TAR qui n'est pas un forma facilement lisible, et elles ne sont pas versionnées, ce qui signifie que n'importe qui peut télécharger une archive correspondant à la dernière version _draft_. 

Je règle cela en mettant le format des archives en ZIP et en les versionnant, tout du moins pour la version publique.
# Notes de QA

Il faut donc cette fois tester que l'archive contient ce qu'elle doit contenir, donc ...
- Créer un article, le publier, puis le modifier sans publier les changements. Télécharger l'archive de la version publique, vérifier que ça correspontd. Télécharger l'archive de la version _draft_ et vérifier que les changement sont bien présent dans cette archive mais pas dans celle de la version publique !!
- Faire de même pour les tutoriels. Pas besoin de sortir la grosse artillerie, un mini-tuto suffira

**Attention** que cette PR doit aussi être testée sous windows, les archives étant normalement directement lisible par cet OS sans besoin d'installer de programme externe.

pierre-24 committed:
ZIP versionnées sur tutos (#1458 et #1460)

pierre-24 committed:
ZIP versionnées sur articles (#1458 et #1460)

pierre-24 committed:
Suppression de l'archive après téléchargement

pierre-24 committed:
Retrait d'une variable inutile

pierre-24 committed:
Retrait d'une variable inutile

firm1 commented on zds/tutorial/views.py:
```diff
@@ -2821,17 +2821,30 @@ def maj_repo_extract(
+    sha = tutorial.sha_draft
+    if 'online' in request.GET and tutorial.sha_public is not None:
+        sha = tutorial.sha_public
+    zip_path = os.path.join('/tmp/',tutorial.slug+'.zip')
```
c'est moi ou tu écris dans le répertoire temporaire du système ici ?

pierre-24 replied:
Ah, juste, crétin que je suis. Je pense que je m'étais posé la question du "ou" et que j'avais pas trouvé de réponse, alors que je l'avais mis là en attendant. T'as pas une idée de ou je pourrais le mettre, du coup ? (le site ne propose pas de répertoire "temporaire" à proprement parler, encore que je peux jouer avec [tempfile](https://docs.python.org/2/library/tempfile.html))

firm1 commented on zds/article/views.py:
```diff
@@ -440,26 +441,30 @@ def maj_repo_article(
+    repo_path = os.path.join(settings.REPO_ARTICLE_PATH, article.get_phy_slug())
+    repo = Repo(repo_path)
+    sha = article.sha_draft
+    if 'online' in request.GET and article.sha_public is not None:
```
la façon dont s'est écrit ici, signifie que l'on peut toujours récupérer l'archive hors ligne. Il suffit de recopier l'url de téléchargement sans la partie `online`

pierre-24 replied:
Bien vu. De manière générale, je vais rajouter des 403 quand on pas les droits d'écriture, comme ce qui se fait pour l'accès au tuto, ça me semble bien.

firm1 replied:
la condition `article.sha_public is not None` ne justifie pas de la mise en ligne d'un article. Il y'a des cas ou on a le `sha = ""` qui peut biaiser ta condition. Il faut donc rajouter (si ce n'est pas encore fait) une méthode dans le model qui fait la même chose que le `on_line` des tutoriels

pierre-24 replied:
J'ai peur que ce soit un peu hors-scope, mais ça ne tiendrais qu'à moi, je ferai un c/c de la méthode load_dic() des tutos qui fait relativement bien ça. Bon, je vais voir ce que je peux faire en ajoutant simplement `on_line` (**EDIT:** qui n'existe pas pour le moment).

pierre-24 replied:
RE-EDIT: tu vas rire : [la fonction on_line() des articles](https://github.com/zestedesavoir/zds-site/blob/dev/zds/article/models.py#L113) fait exactement `is not None`. Je vais rajouter `and is not ""`

firm1 commented on zds/article/views.py:
```diff
@@ -440,26 +441,30 @@ def maj_repo_article(
-        .format(article.slug)
 
+def download(request):
+    """Download a tutorial."""
```
>  `"""Download a tutorial."""`

=> `"""Download an article."""`

pierre-24 replied:
Les joies du copier coller.

firm1 commented on zds/article/views.py:
```diff
@@ -440,26 +441,30 @@ def maj_repo_article(
+    zip_file.close()
+    response = HttpResponse(open(zip_path , "rb").read(), content_type="application/zip")
+    response["Content-Disposition"] = "attachment; filename={0}.zip".format(article.slug)
+    os.remove(zip_path) # TODO: caching (at least for the public version?)
```
`# TODO:` , je pense qu'il faut éviter de rajouter des TODO dans le code, et crée des issues à la place pour les évolutions.

pierre-24 replied:
Ok :)

firm1 commented on zds/tutorial/views.py:
```diff
@@ -2821,17 +2821,30 @@ def maj_repo_extract(
+    repo = Repo(repo_path)
+    sha = tutorial.sha_draft
+    if 'online' in request.GET and tutorial.sha_public is not None:
+        sha = tutorial.sha_public
```
Idem ici, on peut encore télécharger l'archive en rédaction même si c'est déjà publié

pierre-24 committed:
ajout d'un test de sécurité

Seul l'auteur et les personnes possédant le droit
peuvent téléchargeer l'archive contenant la version
draft.

pierre-24 committed:
Correction du dossier temporaire

pierre-24 commented:
J'ai tenu compte de toute les remarques. Comme on dit, 

# QA CAN CONTINUE !

firm1 commented on zds/article/views.py:
```diff
@@ -440,26 +442,33 @@ def maj_repo_article(
+    sha = article.sha_draft
+    if 'online' in request.GET and article.sha_public is not None:
+        sha = article.sha_public
+    elif request.user not in article.authors.all():
```
A vérifier ... que ce passe t-il si on rentre dans cette condition sans être connecté ?

pierre-24 replied:
On se mange bien une erreur 403, ce qui est il me semble le comportement voulu (à noter qu'on peu bel et bien télécharger l'archive de la version _online_, ce qui est, toujours selon moi, le comportement voulu). Je crois que dans ce cas, `request.user` vaut `None`, tout simplement.

pierre-24 committed:
Amélioration de la condition sur la version online

pierre-24 commented:
Et on peut continuer :-)

pierre-24 committed:
Correction d'une permisssion

pierre-24 committed:
Ajout de tests unitaires

pierre-24 committed:
Merge remote-tracking branch 'upstream/dev' into fix_archive_1458_1460

pierre-24 commented:
Hop, j'ai rajouté des tests unitaires qui font bien, donc ...

# QA CAN CONTINUE :)

pierre-24 committed:
Typo

Eskimon commented:
Tout roule chez moi :)
